File: BOJ/[1629] 곱셈/woong-jae/1629.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const input = [];
rl.on('line', function (line) {
    input.push(line.split(" "));
}).on("close", function () {
    console.log(solution(...input[0].map(e => +e)));
    process.exit();
});

function solution(A, B, C) {
    A = BigInt(A);
    C = BigInt(C);
    const dnq = (exponent) => {
        if(exponent === 1) return A % C;

        const half = dnq(Math.floor(exponent / 2)) % C;
        if(exponent % 2 === 0) return half * half % C;
        return half * half * A % C;
    }

    return dnq(B).toString();
}

File: BOJ/[1629] 곱셈/JFe/1629.py
import sys

def calc(a, x):
    # base case
    if x == 1: 
        return a % C
    # 지수를 2씩 나눠서 분할 정복
    temp = calc(a, x // 2)
    # 짝수인 경우
    if x % 2 == 0:
        return temp * temp % C
    # 홀수인 경우
    else: 
        return temp * temp * a % C

A, B, C = map(int, sys.stdin.readline().split())
print(calc(A, B))